pkg/blobinfocache/boltdb/boltdb.go

```diff
@@ -325,7 +325,67 @@ func (bdc *cache) appendReplacementCandidates(candidates []prioritize.CandidateW
 // data from previous RecordDigestUncompressedPair calls is used to also look up variants of the blob which have the same
 // uncompressed digest.
 func (bdc *cache) CandidateLocations2(transport types.ImageTransport, scope types.BICTransportScope, primaryDigest digest.Digest, canSubstitute bool) []blobinfocache.BICReplacementCandidate2 {
-	return bdc.candidateLocations(transport, scope, primaryDigest, canSubstitute, true)
+	resAllBlobs := []prioritize.CandidateWithTime{}
+	var uncompressedDigestValue digest.Digest // = ""
+	requireCompressionInfo := true
+	resultPrioritized := bdc.candidateLocations(transport, scope, primaryDigest, canSubstitute, requireCompressionInfo)
+	// Reprocess all blobs and append them to resultPrioritized in lower priority
+	if err := bdc.view(func(tx *bolt.Tx) error {
+		scopeBucket := tx.Bucket(knownLocationsBucket)
+		if scopeBucket == nil {
+			return nil
+		}
+		scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
+		if scopeBucket == nil {
+			return nil
+		}
+		if err := scopeBucket.ForEach(func(name []byte, v []byte) error {
+			bucket := scopeBucket.Bucket(name)
+			// compressionBucket won't have been created if previous writers never recorded info about compression,
+			// and we don't want to fail just because of that
+			compressionBucket := tx.Bucket(digestCompressorBucket)
+
+			resAllBlobs = append(resAllBlobs, bdc.appendReplacementCandidates(resAllBlobs, bucket, compressionBucket, primaryDigest, requireCompressionInfo)...)
+			if canSubstitute {
+				if uncompressedDigestValue = bdc.uncompressedDigest(tx, primaryDigest); uncompressedDigestValue != "" {
+					b := tx.Bucket(digestByUncompressedBucket)
+					if b != nil {
+						b = b.Bucket([]byte(uncompressedDigestValue.String()))
+						if b != nil {
+							if err := b.ForEach(func(k, _ []byte) error {
+								d, err := digest.Parse(string(k))
+								if err != nil {
+									return err
+								}
+								if d != primaryDigest && d != uncompressedDigestValue {
+									resAllBlobs = append(resAllBlobs, bdc.appendReplacementCandidates(resAllBlobs, bucket, compressionBucket, d, requireCompressionInfo)...)
+								}
+								return nil
+							}); err != nil {
+								return err
+							}
+						}
+					}
+					if uncompressedDigestValue != primaryDigest {
+						resAllBlobs = append(resAllBlobs, bdc.appendReplacementCandidates(resAllBlobs, bucket, compressionBucket, uncompressedDigestValue, requireCompressionInfo)...)
+					}
+				}
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return []blobinfocache.BICReplacementCandidate2{}
+	}
+
+	resultPrioritizedAllBlobs := prioritize.DestructivelyPrioritizeReplacementCandidates(resAllBlobs, primaryDigest, uncompressedDigestValue)
```

Comment:
(The location-unknown digests should probably have their own tunable limit of entries, with a smaller limit.)

Reply:
This can be easily done now, since Candidates contain a new field `UnkownLocation` which should allow to tune early but just curious why is it needed since `LocationUnknown` is already priortized at lowest level in final result and they will get processed at last.

Reply:
All of this is a heuristic. We don’t _know_ that there will be a match. Hypothetically, if we had a 1000 candidates for some weird reason, the user could be waiting for 1000 failing attempts (multiple network roundtrips), all failing, in the worst case.

Limiting the number of returned candidates limits the number of attempts, and thus limits the total downside of the heuristic not finding a match.

Reply:
Ah i agree, i'll make this change.